gh-101178: Add Ascii85, Base85, and Z85 support to binascii

kangtastic opened the pull request:
## Synopsis

Add Ascii85, Base85, and Z85 encoder and decoder functions implemented in C to `binascii` and use them to greatly improve the performance and reduce the memory usage of the existing Ascii85, Base85, and Z85 codec functions in `base64`.

No API or documentation changes are necessary with respect to any functions in `base64`, and all existing unit tests for those functions continue to pass without modification.

Resolves: gh-101178

## Discussion
The base85-related functions in `base64` are now wrappers for the new functions in `binascii`, as envisioned in the [docs](https://docs.python.org/3/library/binascii.html):
> The [binascii](https://docs.python.org/3/library/binascii.html#module-binascii) module contains a number of methods to convert between binary and various ASCII-encoded binary representations. Normally, you will not use these functions directly but use wrapper modules like [uu](https://docs.python.org/3/library/uu.html#module-uu) or [base64](https://docs.python.org/3/library/base64.html#module-base64) instead. The [binascii](https://docs.python.org/3/library/binascii.html#module-binascii) module contains low-level functions written in C for greater speed that are used by the higher-level modules.

Parting out Ascii85 from Base85 and Z85 was warranted in my testing despite the code duplication due to the various performance-murdering special cases in Ascii85.

Comments and questions are welcome.

## Benchmarks

Updated December 28, 2025.

```python
# bench_b85.py

# Note: EXTREMELY SLOW on unmodified mainline CPython
#       when tracing malloc on the base-85 functions.

import base64
import sys
import timeit
import tracemalloc

funcs = [(base64.b64encode, base64.b64decode),  # sanity check/comparison
         (base64.a85encode, base64.a85decode),
         (base64.b85encode, base64.b85decode),
         (base64.z85encode, base64.z85decode)]

def mb(n):
    return f"{n / 1024 / 1024:.3f} MB"

def stats(func, data, t, m):
    name, n, bps = func.__qualname__, len(data), len(data) / t
    print(f"{name} : {n} b in {t:.3f} s ({mb(bps)}/s) using {mb(m)}")

if __name__ == "__main__":
    data = b"a" * int(sys.argv[1]) * 1024 * 1024
    for fenc, fdec in funcs:
        tracemalloc.start()
        enc = fenc(data)
        menc = tracemalloc.get_traced_memory()[1] - len(enc)
        tracemalloc.stop()
        tenc = timeit.timeit("fenc(data)", number=1, globals=globals())
        stats(fenc, data, tenc, menc)

        tracemalloc.start()
        dec = fenc(enc)
        mdec = tracemalloc.get_traced_memory()[1] - len(dec)
        tracemalloc.stop()
        tdec = timeit.timeit("fdec(enc)", number=1, globals=globals())
        stats(fdec, enc, tdec, mdec)
```

```bash
# Python 3.15.0a3+ (heads/main:0efbad60e13, Dec 28 2025, 11:02:16)
# ./configure --enable-optimizations --with-lto

# Unmodified
$ time ./python bench_b85.py 64
b64encode : 67108864 b in 0.092 s (693.266 MB/s) using 42.667 MB
b64decode : 89478488 b in 0.234 s (364.961 MB/s) using 56.889 MB
a85encode : 67108864 b in 7.163 s (8.935 MB/s) using 2664.401 MB
a85decode : 83886080 b in 14.478 s (5.526 MB/s) using 3332.254 MB
b85encode : 67108864 b in 6.965 s (9.189 MB/s) using 2664.401 MB
b85decode : 83886080 b in 10.082 s (7.935 MB/s) using 3332.254 MB
z85encode : 67108864 b in 7.245 s (8.834 MB/s) using 2664.102 MB
z85decode : 83886080 b in 9.666 s (8.277 MB/s) using 3332.254 MB

real    9m44.382s
user    9m27.271s
sys     0m12.747s


# With this PR
b64encode : 67108864 b in 0.085 s (753.375 MB/s) using 42.667 MB
b64decode : 89478488 b in 0.230 s (371.282 MB/s) using 56.889 MB
a85encode : 67108864 b in 0.094 s (681.709 MB/s) using 0.000 MB
a85decode : 83886080 b in 0.191 s (418.019 MB/s) using 0.000 MB
b85encode : 67108864 b in 0.075 s (850.118 MB/s) using 0.000 MB
b85decode : 83886080 b in 0.141 s (567.490 MB/s) using 0.000 MB
z85encode : 67108864 b in 0.074 s (864.559 MB/s) using 0.000 MB
z85decode : 83886080 b in 0.173 s (462.854 MB/s) using 0.000 MB

real    0m1.865s
user    0m1.726s
sys     0m0.126s
```

The old pure-Python implementation is two orders of magnitude slower and uses over O(40n) temporary memory.

kangtastic committed:
Add Ascii85, base85, and Z85 support to binascii

Add Ascii85, base85, and Z85 encoders and decoders to `binascii`,
replacing the existing pure Python implementations in `base64`.

No API or documentation changes are necessary with respect to
`base64.a85encode()`, `b85encode()`, etc., and all existing unit
tests for those functions continue to pass without modification.

Note that attempting to decode Ascii85 or base85 data of length 1 mod 5
(after accounting for Ascii85 quirks) now produces an error, as no
encoder would emit such data. This should be the only significant
externally visible difference compared to the old implementation.

Resolves: gh-101178

AA-Turner commented on Lib/base64.py:
```diff
@@ -290,36 +290,6 @@ def b16decode(s, casefold=False):
-_A85START = b"<~"
-_A85END = b"~>"
-
-def _85encode(b, chars, chars2, pad=False, foldnuls=False, foldspaces=False):
```
Per PEP-0399, the Python implementation must be kept, with the C accelerator and Python implementation tested to ensure they produce identical output.

kangtastic replied:
From PEP-0399:

> If an acceleration module is provided it is to be named the same as the module it is accelerating with an underscore attached as a prefix, e.g., _warnings for warnings. The common pattern to access the accelerated code from the pure Python implementation is to import it with an import *, e.g., from _warnings import *. This is typically done at the end of the module to allow it to overwrite specific Python objects with their accelerated equivalents.

Although the effect is the same, there is a subtle difference in that strictly speaking, this PR isn't providing alternative C implementations for the base 85-related pure-Python functions in `base64`. It's adding new functions into the existing `binascii` C module and turning said Python functions into wrappers for them, which is in keeping with how `binascii` and `base64` have historically been interrelated.

That difference means the guidelines in PEP-0399 don't apply cleanly. So e.g. creating a new `_base64` C module doesn't make sense. Neither does trying to use the accelerated routines only if available, as `binascii` will always be available.

Do you have any thoughts on how to keep the Python implementation in a way that works with Python's import system? I'm not familiar with an analogous situation in the rest of the codebase.

eendebakpt replied:
In https://github.com/python/cpython/pull/91610 a C version is added for deepcopy and unit tests are created for both the c and python implementation. If you search a bit in the codebase you can find some more examples.

kangtastic replied:
Never mind, all. I assumed `binascii`, being part of the stdlib, would/should always be available.

I've since found that [quopri](https://github.com/python/cpython/blob/main/Lib/quopri.py) doesn't make that assumption. I'll do what it does.

kangtastic committed:
Restore base64.py

kangtastic committed:
Create _base64 module with wrappers for accelerated functions

If we were strictly following PEP-0399, _base64 would be a C
module for accelerated functions in base64. Due to historical
reasons, those should actually go in binascii instead.

We still want to preserve the existing Python code in base64.
Parting out facilities for accessing the C functions into a
module named _base64 shouldn't risk a naming conflict and
will simplify testing.

kangtastic committed:
Test both Python and C codepaths in base64

This is done differently to PEP-0399 to minimize the number of
changed lines.

kangtastic committed:
Match behavior between Python and C base 85 functions

As we're now keeping the existing Python base 85 functions, the C
implementations should behave exactly the same, down to exception
type and wording. It is also no longer an error to try to decode
data of length 1 mod 5.

kangtastic committed:
Add Z85 tests to binascii

kangtastic committed:
Update generated files

AA-Turner commented on Lib/_base64.py:
```diff
@@ -0,0 +1,57 @@
+"""C accelerator wrappers for originally pure-Python parts of base64."""
+
+from binascii import Error, a2b_ascii85, a2b_base85, b2a_ascii85, b2a_base85
+from base64 import _bytes_from_decode_data, bytes_types
```
We should avoid import cycles like this, it can make refactoring in the future harder.

AA-Turner commented on Lib/base64.py:
```diff
@@ -576,6 +576,27 @@ def decodebytes(s):
+try:
+    from _base64 import (_a85encode, _a85decode, _b85encode,
+                         _b85decode, _z85encode, _z85decode)
+    from functools import update_wrapper
```
Functools is an expensive import, I would copy the relative parts of `update_wrapper()` locally.

AA-Turner commented on Lib/test/test_base64.py:
```diff
@@ -1,10 +1,37 @@
+c_base64 = import_fresh_module("base64", fresh=["_base64"])
+
+
+def with_c_implementation(test_func):
```
Instead of a decorator, perhaps use the mixin approach from other modules.

kangtastic committed:
Avoid importing functools

Importing update_wrapper() from functools to copy attributes
is expensive. Do it ourselves for only the most relevant ones.

kangtastic committed:
Avoid circular import in _base64

This requires some code duplication, but oh well.

kangtastic committed:
Do not use a decorator for changing exception type

Using a decorator complicates function signature introspection.

kangtastic committed:
Test Python and C codepaths in base64 using mixins

Do we really need to test the legacy API twice?

AA-Turner commented on Lib/base64.py:
```diff
@@ -576,6 +576,27 @@ def decodebytes(s):
+# Use accelerated implementations of originally pure-Python parts if possible.
+try:
+    from _base64 import (_a85encode, _a85decode, _b85encode,
+                         _b85decode, _z85encode, _z85decode)
```
Given these are already in a private module, you can remove the prefix. That means the `_copy_attributes` function only needs to copy `__doc__`, and `__module__` can be set to the static `'base64'`.

kangtastic replied:
Done.

kangtastic committed:
Remove leading underscore from functions in private module

serhiy-storchaka committed:
Merge branch 'main' into gh-101178-rework-base85

serhiy-storchaka committed:
Use more modern C API.

serhiy-storchaka committed:
Fix tests.

serhiy-storchaka committed:
Merge branch 'main' into gh-101178-rework-base85

serhiy-storchaka committed:
Fix new tests.

picnixz commented on Doc/library/binascii.rst:
```diff
@@ -77,6 +77,79 @@ The :mod:`binascii` module defines the following functions:
+   of no more than the specified width separated by the ASCII newline
+   character.
+
+   If *pad* is true, the input is padded to a multiple of 4 before encoding.
```
```suggestion
   If *pad* is true, the input is padded to a multiple of 4 before encoding.

   .. versionadded:: next
```

picnixz commented on Doc/library/binascii.rst:
```diff
@@ -77,6 +77,79 @@ The :mod:`binascii` module defines the following functions:
+   Valid base85 data contains characters from the base85 alphabet in groups
+   of five (except for the final group, which may have from two to five
+   characters). Each group encodes 32 bits of binary data in the range from
+   ``0`` to ``2 ** 32 - 1``, inclusive.
```
```suggestion
   ``0`` to ``2 ** 32 - 1``, inclusive.

   .. versionadded:: next
```

picnixz commented on Doc/library/binascii.rst:
```diff
@@ -77,6 +77,79 @@ The :mod:`binascii` module defines the following functions:
+   *ignore* is an optional bytes-like object that specifies characters to
+   ignore in the input.
+
+   Invalid Ascii85 data will raise :exc:`binascii.Error`.
```
```suggestion
   Invalid Ascii85 data will raise :exc:`binascii.Error`.

   .. versionadded:: next
```

picnixz commented on Doc/library/binascii.rst:
```diff
@@ -77,6 +77,79 @@ The :mod:`binascii` module defines the following functions:
+   If *newline* is true, a newline char is appended to the result.
+
+   If *z85* is true, the Z85 alphabet is used for conversion.
+   See `Z85 specification <https://rfc.zeromq.org/spec/32/>`_ for more information.
```
```suggestion
   See `Z85 specification <https://rfc.zeromq.org/spec/32/>`_ for more information.

   .. versionadded:: next
```

picnixz commented on Lib/_base64.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def _bytes_from_encode_data(b):
+    return b if isinstance(b, bytes_types) else memoryview(b).tobytes()
```
`memoryview(b).tobytes()` may raise a generic TypeError here, whereas it is in `_bytes_from_decode_data`. Should we do it as well?

kangtastic replied:
The [discussion](https://discuss.python.org/t/accelerator-for-ascii85-base85/105415) by @serhiy-storchaka seems to have been resolved with the conclusion that the pure-Python base-85-related functions from `base64` may be removed.

I'll therefore turn `base64.a85encode()`, etc. into wrappers for the new `binascii.b2a_ascii85()`, etc. as I had it originally, similarly to how e.g. `base64.b64encode()` is a wrapper for `binascii.b2a_base64()`.

`_base64.py` will be removed, since it will no longer be necessary to test against both Python and C implementations of the base-85-related functions. So that makes your concerns for this specific file moot.

However, to follow your comment here a bit:

The type conversion logic is replicated from the existing `_85encode()` [helper function](https://github.com/python/cpython/blob/main/Lib/base64.py#L299), which itself seems to have replicated the logic from the existing `_b32encode()` [helper function](https://github.com/python/cpython/blob/main/Lib/base64.py#L163).

The logic in `_b32encode()`, in turn, seems to be an attempt to replicate the existing `binascii` [facility](https://github.com/python/cpython/blob/main/Modules/binascii.c#L164) for input type conversions.

The new additions (`base64.a85encode()`, etc.) use the existing `binascii` facility, which raises a more descriptive `TypeError` instead of a generic one.

Can you or someone else verify that the C type conversion logic in `binascii` and the Python logic in `base64` are equivalent, or at least equivalent enough?

picnixz replied:
Oh so it was something that was already there. Ok, sorry for the noise. I was more worried about the fact that a TypeError was caught during decoding but not during encoding.

picnixz commented on Lib/_base64.py:
```diff
@@ -0,0 +1,79 @@
+
+# Functions in binascii raise binascii.Error instead of ValueError.
+
+def a85encode(b, *, foldspaces=False, wrapcol=0, pad=False, adobe=False):
```
```suggestion
def a85encode(b, /, *, foldspaces=False, wrapcol=0, pad=False, adobe=False):
```

serhiy-storchaka replied:
No, the existing Python implementation does not use positional-only parameters. Changing this is a separate issue.

picnixz replied:
Arf, ok :( I thought it was incorrectly c/c.

picnixz commented on Lib/_base64.py:
```diff
@@ -0,0 +1,79 @@
+        raise ValueError(e) from None
+
+
+def a85decode(b, *, foldspaces=False, adobe=False, ignorechars=b' \t\n\r\v'):
```
```suggestion
def a85decode(b, /, *, foldspaces=False, adobe=False, ignorechars=b' \t\n\r\v'):
```

picnixz commented on Modules/binascii.c:
```diff
@@ -592,6 +670,442 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+        if (this_digit < 85) {
+            if (leftchar > UINT32_MAX / 85 ||
+                (leftchar *= 85) > UINT32_MAX - this_digit) {
+                state = get_binascii_state(module);
+                if (state != NULL) {
+                    PyErr_SetString(state->Error, "Ascii85 overflow");
+                }
```
This can be NULL only if:

- the state is really missing (`mod->md_state == NULL`)
- 'module' is not a module

I don't know when it could be possible for the state to be missing (maybe at interpreter's finalizaation?).

kangtastic replied:
> I don't know when it could be possible for the state to be missing (maybe at interpreter's finalizaation?).

Apparently it's possible for state to be missing [if an error is already set](https://github.com/python/cpython/blob/main/Modules/binascii.c#L491).

picnixz replied:
Yeah but this is only if the error is a badargument (that is we didn't give a module).

kangtastic replied:
I think I did it because every other `PyErr_SetString(state->)` or `PyErr_Format(state->)` in `binascii` does it :sweat_smile:

Maybe someone knows if it's OK to do `assert(state != NULL);` instead, like some other modules. Separate issue?

picnixz replied:
Yeah I think it's better to do it in a separate issue.

picnixz commented on Modules/binascii.c:
```diff
@@ -592,6 +670,442 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+
+    return PyBytesWriter_FinishWithPointer(writer, bin_data);
+
+error_end:
```
I'd suggest naming this `error` instead of `error_end`.

kangtastic replied:
Will do.

picnixz reviewed (commented):
A few comments. I'd appreciate, as it's new code, that PEP-7 is followed.

serhiy-storchaka committed:
Optimize binascii.b2a_ascii85().

serhiy-storchaka committed:
Apply suggestions from code review

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

kangtastic committed:
Update C style to more closely adhere to PEP-7

kangtastic committed:
Remove pure-Python base-85-related codepaths in base64

Per [1] this will be allowed despite PEP-0399.

[1]: https://discuss.python.org/t/accelerator-for-ascii85-base85/105415/3

kangtastic committed:
Remove now-unnecessary _base64 module and fix tests

We are no longer testing both Python and C codepaths for
base-85-related functions in base64.

kangtastic committed:
Separate Z85 from Base85 on the Python API side

This avoids exposing a `z85` bool parameter.

Also update documentation. While we're at it, refer to the
encoding as "Base85" instead of "base85".

kangtastic committed:
Fix tests after separating Base85 from Z85

kangtastic committed:
Merge branch 'main' into gh-101178-rework-base85

kangtastic committed:
Update generated files after merging main

kangtastic committed:
Update Misc/NEWS.d and Misc/ACKS

kangtastic committed:
Update generated files again

kangtastic committed:
Fix typo in NEWS entry

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+    wrap: bool = False
+        Expect data to be wrapped in '<~' and '~>' as in Adobe Ascii85.
+    ignore: Py_buffer(c_default="NULL", py_default="b''") = None
+        An optional bytes-like object with input characters to be ignored.
```
I have a question -- why do you use different parameter names than in `base64.a85encode()`?

kangtastic replied:
I just chose some names that sounded OK to me. I didn't like `adobe` in `base64.a85decode()` because it seemed too specific, and there wasn't commonality between `binascii` and `base64` parameter names (lots of legacy code everywhere).

After reading #143262 and #143216 I gather that reusing CPython interned strings is implicitly good practice, so I will try to do that if you prefer.

serhiy-storchaka replied:
I think that it would be less confusing if both interfaces, in `binascii` and `base64`, would use the same names for parameters, if possible. I don't know why `strict_mode` and `validate` differ, this is confusing.

"ignore" was previously only used as the name of the error handler (like "strict", "replace", etc).

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+        ascii_len -= 2;
+        if (ascii_len >= 2
+            && ascii_data[0] == BASE85_A85_PREFIX
+            && ascii_data[1] == BASE85_A85_AFFIX) {
```
PEP 7: if the condition takes multiple lines, move `{` at a separate line.
```suggestion
            && ascii_data[1] == BASE85_A85_AFFIX)
        {
```

kangtastic replied:
Thanks, missed this, will do.

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+            bin_len += 4;
+        }
+    }
+    bin_len = 4 * ((bin_len + 4) / 5);
```
This calculation is not accurate, because `z`s and `y`s already accounted in the total number of characters. It is also possible to get an integer overflow on 32-bit platform -- try to decode a string of `2**29` `z`s.

Count separately the number of `z`s and `y`s, then the output size is  `(bin_len - count + 4) / 5 * 4 + count * 4`, but check for integer overflow.

kangtastic replied:
Could you explain what you mean? `z` and `y` are respectively equivalent to `!!!!!` and `+<VdL`, so it's correct to add 4 to `ascii_len` for each one encountered.

To be thorough I wrote the following to investigate if there are differences between the current and proposed output length calculation. It prints `diffs: 0`.

```c
#include <stdio.h>

int main() {
  int diffs = 0;

  for (int ascii_len = 1; ascii_len <= 1024; ascii_len++) {
    for (int count_yz = 0; count_yz <= ascii_len; count_yz++) {
      /* Current output length calculation */
      int adj_ascii_len = ascii_len + 4 * count_yz;
      int bin_len = 4 * ((adj_ascii_len + 4) / 5);

      /* Proposed output length calculation */
      int bin_len_2 = (ascii_len - count_yz + 4) / 5 * 4 + count_yz * 4;

      if (bin_len != bin_len_2) {
        printf("ascii_len %d count_yz %d bin_len %d bin_len_2 %d\n",
          ascii_len, count_yz, bin_len, bin_len_2);
        diffs++;
      }
    }
  }

  printf("diffs: %d\n", diffs);

  return 0;
}
```

I didn't check for integer overflow because the rest of `binascii` doesn't, but I can add it.

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+        Allow 'y' as a short form encoding four spaces.
+    wrap: bool = False
+        Expect data to be wrapped in '<~' and '~>' as in Adobe Ascii85.
+    ignore: Py_buffer(c_default="NULL", py_default="b''") = None
```
`None` is not acceptable value, isn't? Use `b''`.

kangtastic replied:
If the signature of `a2b_ascii85(string, /, *, fold_spaces=False, wrap=False, ignore=b"")` was `ignore=None`, this would be acceptable. Because it isn't, this clinic input should be `ignore: Py_buffer(c_default="NULL", py_default="b''") = b''` instead. Is that what you meant?

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+        Emit 'y' as a short form encoding four spaces.
+    wrap: bool = False
+        Wrap result in '<~' and '~>' as in Adobe Ascii85.
+    width: unsigned_int(bitwise=True) = 0
```
`bitwise=True` make it to wrap, so `2**32` is interpreted as 0.

There is also no benefit of restricting the range to the C's int, it have to support `sys.maxsize`. So use `size_t` here. We can also use a special converter, because the current Python code does not have limitation, and negative values interpreted as 1 (although this is an implementation detail, not documented and not tested, so it is not necessary to support this).

kangtastic replied:
I'll investigate this further and either use size_t or a custom converter.

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+    }
+
+    /* Allocate output buffer.
+       XXX: Do a pre-pass above some threshold estimate (cf. 'yz')?
```
I do not think it is worth it.

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+    data: ascii_buffer
+    /
+    *
+    strict_mode: bool = False
```
Do we need this? It is always True in `base64`. `strict_mode=False` is legacy.

kangtastic replied:
Maybe not so legacy, a `strict_mode` parameter was added to `binascii.a2b_base64()` in #24402 for what seem like good security-related reasons that are possibly relevant for `binascii.a2b_base85()` and `binascii.a2b_z85()` as well.

serhiy-storchaka replied:
Base64 decoder works in non-strict mode by default. This is legacy. The `strict_mode` parameter was added to change the default behavior. But Base85 decoder never worked in non-strict mode and does not need to.

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -592,6 +692,511 @@ binascii_b2a_base64_impl(PyObject *module, Py_buffer *data, int newline)
+    *
+    pad: bool = False
+        Pad input to a multiple of 4 before encoding.
+    newline: bool = True
```
Do we need this? It is always False in `base64`. `newline=True` is legacy.

kangtastic replied:
For this one I agree we probably don't really need it. I put it in following the example of `binascii.b2a_base64()`. Note that `binascii.b2a_uu()` always appends a newline but `binascii.b2a_hex()` doesn't. If you prefer I can remove it.

serhiy-storchaka replied:
`b2a_base64()` adds a newline by default. The `newline` parameter was added to alter the default behavior. But `b2a_base85()` does not have such historical baggage, it should not add a newline by default, and does not need to add it optionally (you always can add it after calling the function).

serhiy-storchaka reviewed (commented):
Fantastic!

I will make final reviews after merging #143262 and #143216 (and maybe some other PRs). Some code, docs and tests can be rewritten after them. Here are some general comments.

serhiy-storchaka committed:
Merge branch 'main' into gh-101178-rework-base85

serhiy-storchaka committed:
Move the NEWS entry to the correct section.

kangtastic committed:
Minor cleanups, align lookup tables to 64 bytes (NFC)

No performance changes. Base85 tables were probably already aligned.

kangtastic committed:
Allow up to sys.maxsize output length when encoding base 85

kangtastic committed:
Fix Ascii85 test from mainline

kangtastic committed:
Allow up to sys.maxsize output length when decoding base 85

Include an integer overflow check for Ascii85.

kangtastic committed:
Defer base 85 overflow check during decoding

Performance gains of up to 8% for a2b_ascii85() and 25% for a2b_base85()
and a2b_z85() were observed.

kangtastic committed:
Merge branch 'main' into gh-101178-rework-base85

serhiy-storchaka commented on Modules/binascii.c:
```diff
@@ -736,6 +814,539 @@ toolong:;
+    }
+
+    /* Build ignore map. */
+    unsigned char ignore_map[256] = {0};
```
This may be slow. Look how it was done for the Base64 decoder.

serhiy-storchaka reviewed (commented):
Thank you for your update @kangtastic. I added support of `ignorechars` in the Base64 decoder, so now the similar code can be used in the Ascii85 decoder. Also note the changes in the Ascii85 documentation.

* Use the same names for parameters in `binascii` and `base64`.
* Remove unneeded parameters.
* Reuse the `ignorechar()` function for `ignorechars`.
* Update the Ascii85 documentation -- refer to "bytes-like object" for "ignorechars". We can now guarantee this.

serhiy-storchaka committed:
Merge branch 'main' into gh-101178-rework-base85

serhiy-storchaka committed:
Rename parameters to match the base64 module.

serhiy-storchaka committed:
Remove parameters strict_mode and newline.

serhiy-storchaka committed:
Optimize ignorechars cache.

serhiy-storchaka committed:
Harmonize documentation.

serhiy-storchaka committed:
Add What's New entries.

serhiy-storchaka committed:
Polish tests.

kangtastic committed:
Rename internal Base 85 codec functions to match Base 64 helpers
